### ISSUE TITLE
Access framework HTTP Request from method

### ISSUE THREAD

**Issue by john-g-g** [bcb/jsonrpcserver#44]
This is a question as much as an issue.  I have several jsonrpc methods that require db, redis, and other state/config access that would normally be passed using the the 'app' or 'request' object. What is the best method to access the parent http request and/or web app object from a method?

It seems like its necessary to modify/create/subclass:
- [aio.methods](https://github.com/bcb/jsonrpcserver/blob/7dd5dfffa210af6e6e868093523f4ac8e3bbe2e4/jsonrpcserver/aio.py#L6)
- [async_methods.py](https://github.com/bcb/jsonrpcserver/blob/master/jsonrpcserver/async_methods.py)
- [async_dispatcher.dispatch](https://github.com/bcb/jsonrpcserver/blob/master/jsonrpcserver/async_dispatcher.py#L43)
- [AsyncRequest](https://github.com/bcb/jsonrpcserver/blob/master/jsonrpcserver/async_request.py)

I will be making these changes for use in our projects ([jussi](https://github.com/steemit/jussi),[yo](https://github.com/steemit/yo),[sbds](https://github.com/steemit/sbds),[hivemind](https://github.com/steemit/hivemind)). 

I'm thinking that changing the signature of the dispatch method (and all the other methods that lead to the jsonrpc method) to also accept `*args, **kwargs` is the best, most modular way:
```python

async def handle(aio_http_request):
    statsd = aio_http_request.app.config['statsd_client']
    jsonrpc_request = await aio_http_request.text()
    response = await methods.dispatch(jsonrpc_request, aio_http_request, statsd=statsd)
    return web.json_response(response)

```
I presume the use case for db/cache/metrics isn't rare, so I thought I would solicit your expertise and opinion about how these changes are implemented, especially if you're interested in adding my changes upstream. 

Thanks for you excellent work!

**Comment by bcb** [comment]
Hi @john-g-g,

If you were using aiohttp's class-based views, you'd get the aiohttp context within your jsonrpc method:
```python
class MyView(web.View):
    @methods.add
    async def ping(self):
        aio_http_request = self.request
        statsd = self.request.app.config['statsd_client']
        # ...

    async def post(self):
        request = await self.request.text()
        response = await methods.dispatch(request)
        return web.json_response(response)

app.router.add_route('*', '/', MyView)
```
However I've just tried this and it doesn't work right now, but if I fix that would it suit your situation?

**Comment by bcb** [comment]
Failing that your idea sounds good, my only suggestion would be to pass the values in an object like:
```python
context = {'request': aio_http_request, 'statsd': statsd}
response = await methods.dispatch(jsonrpc_request, context=context)
```
And methods would receive the context object:
```python
@methods.add
async def ping(context):
    statsd = context['statsd']
    # ...
```

**Comment by bcb** [comment]
I was able to access the aiohttp values in my method like this:

```python
from aiohttp import web
from jsonrpcserver.aio import dispatch
from jsonrpcserver.response import NotificationResponse

class MyView(web.View):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.methods = (self.ping,)

    async def ping(self):
        request = self.request
        # ...
        return 'pong'

    async def post(self):
        request = await self.request.text()
        response = await dispatch(self.methods, request)
        if isinstance(response, NotificationResponse):
            return web.Response()
        else:
            return web.json_response(response)

app = web.Application()
app.router.add_route('*', '/', MyView)

if __name__ == '__main__':
    web.run_app(app, port=5000)
```

**Comment by semirook** [comment]
The best I can do currently is this workaround:

```python
async def rpc_handler(request):
    raw_request = await request.text()
    json_request = json.loads(raw_request)
    json_request.setdefault('params', {}).update({'request': request})

    response = await methods.dispatch(json_request)

    return web.json_response(response)
```

And then in every method I have access to the original request object. It works but I don't like it.

```python
@methods.add
async def my_method(request):
    pass
```

**Comment by bcb** [comment]
Ok, I'll add the ability to pass extra data through `dispatch`.

**Comment by bcb** [comment]
I'm at Pycon this weekend, but I will get to this next week

**Comment by semirook** [comment]
Thanks a lot!

**Comment by bcb** [comment]
Ok, I've released a new version (3.5.0) which allows you to pass context through to the methods.

Here's how your code would look @semirook 

```python
@methods.add
async def my_method(context=None):
    # Use request here...
    pass

async def rpc_handler(request):
    raw_request = await request.text()
    response = await methods.dispatch(raw_request, context=request)
    if response.is_notification:
        return web.Response()
    else:
        return web.json_response(response)
```

Still, class-based views are a better option, if you can use them. (see [above](https://github.com/bcb/jsonrpcserver/issues/44#issuecomment-314696598))
